File: 583-delete-operation-for-two-strings/delete-operation-for-two-strings.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>> dp(2,
                               vector<int>(word2.length() + 1, -1));
        for (int i=0; i<=word1.length(); i+=1){
            bool curr = i & 1;
            for (int j=0; j<=word2.length(); j+=1){
                if (i == 0 || j == 0){
                    dp[curr][j] = 0;
                } else if (word1[i-1] == word2[j-1]){
                    dp[curr][j] = 1 + dp[1-curr][j-1];
                } else {
                    dp[curr][j] = max(dp[1-curr][j], dp[curr][j-1]);
                }
            }
        }

        return (word1.length() + word2.length()) - (2 * dp[word1.length() & 1][word2.length()]);
    }
};

File: 583-delete-operation-for-two-strings/delete-operation-for-two-strings.go
func minDistance(word1 string, word2 string) int {
    n, m := len(word1), len(word2)
    dp := make([]int, m+1)
    for i := 1; i <= n; i++ {
        temp := make([]int, m+1)
        for j := 1; j <= m; j++ {
            if word1[i-1] == word2[j-1] {
                temp[j] = 1 + dp[j-1]
            } else {
                temp[j] = max(dp[j], temp[j-1])
            }
        }
        dp = temp
    }

    return (n + m - (2 * dp[m]))
}